From: http://xenbits.xensource.com/linux-2.6.18-xen.hg?rev/c0f2f398aa3c
# HG changeset 832 patch
# User Keir Fraser <keir.fraser@citrix.com>
# Date 1237377065 0
# Node ID c0f2f398aa3cab195b632531447e95110e9d8282
# Parent  c1f0373ff44ea3f9ba3845e05faf3d27d7d52731
Subject: x86: Fix interaction of NTP and dom0->xen time updates

Don't discard NTP sync when updating Xen wallclock time from dom0,
as that's almost the first thing we do when we become synced.
Move the call to ntp_clear() into do_settimeofday(), which is the
only caller of __update_wallclock() that looks like it should break
NTP sync.

This fixes the timer chain that sets Xen's wallclock every minute when
dom0 is NTP synced, which in turn greatly improves wallclock accuracy
in PV domU.

Signed-off-by: Tim Deegan <Tim.Deegan@citrix.com>
Acked-by: jbeulich@novell.com

--- sle11-2009-03-24.orig/arch/x86/kernel/time_32-xen.c	2009-03-24 10:00:15.000000000 +0100
+++ sle11-2009-03-24/arch/x86/kernel/time_32-xen.c	2009-03-18 12:51:05.000000000 +0100
@@ -286,8 +286,6 @@ static void __update_wallclock(time_t se
 
 	set_normalized_timespec(&xtime, xtime_sec, xtime_nsec);
 	set_normalized_timespec(&wall_to_monotonic, wtm_sec, wtm_nsec);
-
-	ntp_clear();
 }
 
 static void update_wallclock(void)
@@ -496,6 +494,7 @@ int do_settimeofday(struct timespec *tv)
 		__normalize_time(&sec, &nsec);
 		__update_wallclock(sec, nsec);
 	}
+	ntp_clear();
 
 	/* Reset monotonic gettimeofday() timeval. */
 	spin_lock(&monotonic_lock);
